drivers/i2c/i2c_renesas_ra_sci.c

```diff
@@ -0,0 +1,731 @@
+			 */
+			if (OPERATION(current) != OPERATION(next)) {
+				if (!(next->flags & I2C_MSG_RESTART)) {
+					LOG_ERR("Restart condition between messages of "
+						"different directions is required."
+						"Current/Total: [%d/%d]",
+						i, num_msgs);
+					ret = -EIO;
+					break;
+				}
+			}
+
+			/* Stop condition is only allowed on last message */
+			if (current->flags & I2C_MSG_STOP) {
+				LOG_ERR("Invalid stop flag. Stop condition is only allowed on "
+					"last message. "
+					"Current/Total: [%d/%d]",
+					i, num_msgs);
+				ret = -EIO;
+				break;
+			}
+		} else {
+			current->flags |= I2C_MSG_STOP;
+		}
+
+		current++;
+	}
+
+	if (ret) {
+		return ret;
+	}
+
+	k_sem_take(&data->bus_lock, K_FOREVER);
+
+	/* Set destination address with configured address mode before sending msg. */
+
+	i2c_master_addr_mode_t addr_mode = 0;
+
+	if (I2C_MSG_ADDR_10_BITS & data->dev_config) {
+		addr_mode = I2C_MASTER_ADDR_MODE_10BIT;
+	} else {
+		addr_mode = I2C_MASTER_ADDR_MODE_7BIT;
+	}
+
+	R_SCI_I2C_SlaveAddressSet(&data->ctrl, addr, addr_mode);
+
+	/* Process input `msgs`. */
+
+	current = msgs;
+
+	while (num_msgs > 0) {
+		if (num_msgs > 1) {
+			next = current + 1;
+		} else {
+			next = NULL;
+		}
+
+		if (current->flags & I2C_MSG_READ) {
+			fsp_err = R_SCI_I2C_Read(&data->ctrl, current->buf, current->len,
+						   next != NULL && (next->flags & I2C_MSG_RESTART));
+		} else {
+			fsp_err =
+				R_SCI_I2C_Write(&data->ctrl, current->buf, current->len,
+						  next != NULL && (next->flags & I2C_MSG_RESTART));
+		}
+
+		if (fsp_err != FSP_SUCCESS) {
+			switch (fsp_err) {
+			case FSP_ERR_INVALID_SIZE:
+				LOG_ERR("Provided number of bytes more than uint16_t size "
+					"(65535) while DTC is used for data transfer.");
+				break;
+			case FSP_ERR_IN_USE:
+				LOG_ERR("Bus busy condition. Another transfer was in progress.");
+				break;
+			default:
+				LOG_ERR("Unknown error.");
+				break;
+			}
+
+			ret = -EIO;
+			goto RELEASE_BUS;
+		}
+
+		/* Wait for callback to return. */
+		k_sem_take(&data->complete_sem, K_FOREVER);
+
+		/* Handle event msg from callback. */
+		switch (data->event) {
+		case I2C_MASTER_EVENT_ABORTED:
+			LOG_ERR("%s failed.", (current->flags & I2C_MSG_READ) ? "Read" : "Write");
+			ret = -EIO;
+			goto RELEASE_BUS;
+		case I2C_MASTER_EVENT_RX_COMPLETE:
+			break;
+		case I2C_MASTER_EVENT_TX_COMPLETE:
+			break;
+		default:
+			break;
+		}
+
+		current++;
+		num_msgs--;
+	}
+
+RELEASE_BUS:
+	k_sem_give(&data->bus_lock);
+
+	return ret;
+}
+
+#ifdef CONFIG_I2C_CALLBACK
+
+static void renesas_ra_sci_i2c_async_done(const struct device *dev, struct sci_i2c_data *data,
+					    int result)
+{
+
+	i2c_callback_t cb = data->cb;
+	void *p_context = data->p_context;
+
+	data->msg_idx = 0;
+	data->msgs = NULL;
+	data->num_msgs = 0;
+	data->cb = NULL;
+	data->p_context = NULL;
+	data->addr = 0;
+
+	k_sem_give(&data->bus_lock);
+
+	/* Callback may wish to start another transfer */
+	cb(dev, result, p_context);
```

Comment:
Better to have a NULL check
```suggestion
    if (cb) {
        cb(dev, result, p_context);
    }
```